models/user_setting.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"strings"
+
+	"code.gitea.io/gitea/models/db"
+
+	"xorm.io/builder"
+)
+
+// UserSetting is a key value store of user settings
+type UserSetting struct {
+	ID     int64  `xorm:"pk autoincr"`
+	UserID int64  `xorm:"index unique(key_userid)"`              // to load all of someone's settings
+	Key    string `xorm:"varchar(255) index unique(key_userid)"` // ensure key is always lowercase
+	Value  string `xorm:"text"`
+}
+
+// BeforeInsert will be invoked by XORM before inserting a record
+func (setting *UserSetting) BeforeInsert() {
+	setting.Key = strings.ToLower(setting.Key)
+}
+
+// BeforeUpdate will be invoked by XORM before updating a record
+func (setting *UserSetting) BeforeUpdate() {
+	setting.Key = strings.ToLower(setting.Key)
+}
+
+func init() {
+	db.RegisterModel(new(UserSetting))
+}
+
+// GetUserSetting returns specific settings from user
+func GetUserSetting(uid int64, keys []string) ([]*UserSetting, error) {
+	settings := make([]*UserSetting, 0, 5)
```

Comment:
Shouldn't it be
```suggestion
	settings := make([]*UserSetting, 0, len(keys))
```
as we expect an entry for each provided key?